--- ncbiquery_orig.py	2017-12-14 10:27:48.419784630 +0100
+++ ncbiquery.py	2017-12-14 10:26:08.249787078 +0100
@@ -641,7 +641,7 @@
     if not targz_file:
         import urllib
         print >>sys.stderr, 'Downloading taxdump.tar.gz from NCBI FTP site...'
-        urllib.urlretrieve("ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz", "taxdump.tar.gz")
+        urllib.urlretrieve("https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz", "taxdump.tar.gz")
         print >>sys.stderr, 'Done. Parsing...'
         targz_file = "taxdump.tar.gz"
 
